File: fill-polygon-impl/src/main/java/com/orangebyte256/convexcopier/fillpolygonimpl/FillPolygonImpl.java
package com.orangebyte256.convexcopier.fillpolygonimpl;

import com.orangebyte256.convexcopier.common.Polygon;
import com.orangebyte256.convexcopier.common.Line;
import com.orangebyte256.convexcopier.common.Point;

import java.util.*;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;

public class FillPolygonImpl {
    private final AtomicInteger curY;
    private final int[] imagePixels, patternPixels;
    private final int imageWidth, patternWidth;

    public FillPolygonImpl(int[] imagePixels, int imageWidth, int[] patternPixels, int patternWidth) {
        this.imagePixels = imagePixels;
        this.patternPixels = patternPixels;
        this.imageWidth = imageWidth;
        this.patternWidth = patternWidth;
        curY = new AtomicInteger();
    }

    protected static int calcOffset(int x, int y, int width) {
        return y * width + x;
    }

    protected static List<Integer> getCrossingPoint(HashSet<Line> crossingSet, int y) {
        return crossingSet.stream().map(l -> l.getXByY(y).get()).sorted().toList();
    }

    private void fillPolygonWorker(HashSet<Line> crossingSet, HashMap<Integer, ArrayList<Line>> linesPerHorizonUpperPoint,
                       HashMap<Integer, ArrayList<Line>> linesPerHorizonBottomPoint, int lastY, Point anchor) {
        while (true) {
            final List<Integer> crossPoints;
            final int y;
            synchronized (curY) {
                y = curY.getAndDecrement();
                if (y <= lastY) {
                    return;
                }
                if (linesPerHorizonUpperPoint.containsKey(y)) {
                    crossingSet.addAll(linesPerHorizonUpperPoint.get(y));
                }
                if (linesPerHorizonBottomPoint.containsKey(y)) {
                    crossingSet.removeAll(linesPerHorizonBottomPoint.get(y));
                }
                crossPoints = getCrossingPoint(crossingSet, y);
                assert (crossPoints.size() % 2) == 0;
            }

            Iterator<Integer> iter = crossPoints.iterator();
            while (iter.hasNext()) {
                int left = iter.next();
                int right = iter.next();
                int length = right - left;
                if (length > 0) {
                    System.arraycopy(patternPixels, calcOffset(left, y, patternWidth), imagePixels,
                            calcOffset(left + anchor.x, y + anchor.y, imageWidth), length);
                }
            }
        }
    }

    // setup all maps which allows to support actual set for checking of crossing
    protected static void fillLinesPerHorizonMaps(Polygon polygon, HashMap<Integer, ArrayList<Line>> linesPerHorizonUpperPoint,
                                           HashMap<Integer, ArrayList<Line>> linesPerHorizonBottomPoint) {
        polygon.getLines().forEach(line -> {
            BiConsumer<HashMap<Integer, ArrayList<Line>>, Integer> addValueToMap = (map, y) -> {
                map.putIfAbsent(y, new ArrayList<>());
                map.get(y).add(line);
            };
            Point first = line.getFirst(), second = line.getSecond();
            addValueToMap.accept(linesPerHorizonUpperPoint, Math.max(first.y, second.y));
            addValueToMap.accept(linesPerHorizonBottomPoint, Math.min(first.y, second.y));
        });
    }

    // Basic algorithm of filling polygon:
    // We are moving horizontal line between the most top and bottom point of polygon.
    // For each horizontal line we start moving from left to right and count each intersection.
    // If amount of intersection is odd we could start copying pixels from pattern until amount not became even
    // Improvement:
    // Instead of checking crossing each lines, we could check only lines which definitely have intersection.
    // For each horizontal line we support actual set for checking of crossing for this purpose.
    // Current version of algorithm supports multithreading where each thread processing another horizontal lines.
    public void fillPolygon(Polygon polygon, int parallelism, Point anchor) {
        HashMap<Integer, ArrayList<Line>> linesPerHorizonUpperPoint = new HashMap<>();
        HashMap<Integer, ArrayList<Line>> linesPerHorizonBottomPoint = new HashMap<>();
        HashSet<Line> crossingSet = new HashSet<>();

        fillLinesPerHorizonMaps(polygon, linesPerHorizonUpperPoint, linesPerHorizonBottomPoint);
        Point enclosingMinPoint = polygon.enclosingMinPoint();
        Point enclosingMaxPoint = polygon.enclosingMaxPoint();
        curY.set(enclosingMaxPoint.y);

        Runnable worker = () -> fillPolygonWorker(crossingSet, linesPerHorizonUpperPoint, linesPerHorizonBottomPoint,
                enclosingMinPoint.y, anchor);
        try {
            Thread[] threads = new Thread[parallelism];
            for (int i = 0; i < parallelism; i++) {
                threads[i] = new Thread(worker);
                threads[i].start();
            }
            for (Thread thread : threads) {
                thread.join();
            }
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: fill-polygon-impl-cpp/FillPolygonImpl.cpp
#include "FillPolygonImpl.h"

#include <set>
#include <vector>
#include <thread>

FillPolygonImpl::FillPolygonImpl(int *imagePixels, int imageWidth, int *patternPixels, int patternWidth) {
    this->imagePixels = imagePixels;
    this->patternPixels = patternPixels;
    this->imageWidth = imageWidth;
    this->patternWidth = patternWidth;
}

std::set<int> FillPolygonImpl::calcCrossingPoints(int y) const {
    std::set<int> crossPoints;
    for (const Line& elem: crossingSet) {
        int x = elem.getXByY(y);
        if (crossPoints.find(x) != crossPoints.end()) {
            crossPoints.erase(x);
        } else {
            crossPoints.insert(x);
        }
    }
    assert (crossPoints.size() % 2 == 0);
    return crossPoints;
}

int FillPolygonImpl::calcOffsetInImage(int x, int y, int width) const {
    return y * width + x;
}

void FillPolygonImpl::fillPolygonWorker(int ancorX, int ancorY) {
    while (true) {
        mutex.lock();
        int y = curY--;
        if (y <= lastY) {
            mutex.unlock();
            return;
        }
        if (linesPerHorizonUpperPoint.find(y) != linesPerHorizonUpperPoint.end()) {
            for (const Line &elem: linesPerHorizonUpperPoint[y]) {
                crossingSet.insert(elem);
            }
        }
        if (linesPerHorizonBottomPoint.find(y) != linesPerHorizonBottomPoint.end()) {
            for (const Line &elem: linesPerHorizonBottomPoint[y]) {
                crossingSet.erase(elem);
            }
        }
        std::set<int> crossPoints = calcCrossingPoints(y);
        mutex.unlock();

        auto orderedPoints = crossPoints.begin();
        while (orderedPoints != crossPoints.end()) {
            int left = *orderedPoints++;
            int right = *orderedPoints++;
            int length = right - left;
            if (length > 0) {
                std::memcpy(imagePixels + calcOffsetInImage(left + ancorX, y + ancorY, imageWidth),
                            patternPixels + calcOffsetInImage(left, y, patternWidth), length * sizeof(int));
            }
        }
    }
}

void FillPolygonImpl::fillPolygon(int coordsSize, int* coordsArray, int parallelism, int ancorX, int ancorY) {
    int minY = INT_MAX, maxY = INT_MIN;
    int pointsCount = coordsSize / 2;
    for (int i = 0; i < pointsCount; i++) {
        Point cur(coordsArray + i * 2);
        Point next(coordsArray + ((i + 1) % pointsCount) * 2);
        minY = std::min(minY, cur.y);
        maxY = std::max(maxY, cur.y);
        Point upper = cur;
        Point bottom = next;
        if (upper.y < bottom.y) {
            std::swap(upper, bottom);
        }
        Line line = Line(upper, bottom);
        linesPerHorizonUpperPoint[upper.y].push_back(line);
        linesPerHorizonBottomPoint[bottom.y].push_back(line);
    }
    curY = maxY;
    lastY = minY;

    std::thread threads[parallelism];
    for (int i = 0; i < parallelism; i++) {
        threads[i] = std::thread(&FillPolygonImpl::fillPolygonWorker, this, ancorX, ancorY);
    }
    for (int i = 0; i < parallelism; i++) {
        threads[i].join();
    }
}